geco/mips/facility_location/orlib.py

```diff
@@ -0,0 +1,134 @@
+import pyscipopt as scip
+import geco.mips.loading.orlib as orlib
+from geco.mips.facility_location.generic import capacitated_warehouse_location
+
+
+def _cap_numeric_reader(file):
+    """
+    Reads cap(NUMBER) Capacitated Warehouse Location instances mentioned in [1].
+
+    Parameters
+    ----------
+    file: file-like object
+
+    Returns
+    -------
+    model: scip.Model
+        A pyscipopt model of the loaded instance
+
+    References
+    ----------
+    ..[1] http://people.brunel.ac.uk/~mastjjb/jeb/orlib/capinfo.html
+    """
+    num_of_warehouses, num_of_customers = orlib.read_numbers(file.readline())
+    capacities = []
+    fixed_costs = []
+    demands = []
+    allocation_cost_per_warehouse = {}
+
+    for _ in range(num_of_warehouses):
+        capacity, fixed_cost = orlib.read_numbers(file.readline())
+        capacities.append(capacity)
+        fixed_costs.append(fixed_cost)
+
+    for i in range(num_of_customers):
+        demand = orlib.read_number(file.readline())
+        allocation_costs = orlib.read_multiline_numbers(file, num_of_warehouses)
+        demands.append(demand)
+        for j, cost in enumerate(allocation_costs):
+            allocation_cost_per_warehouse[i, j] = cost
+
+    return capacitated_warehouse_location(
```

Comment:
I think we should return all parameters and leave it to the user which formulation to pass them to.

Reply:
How about the current solution? I separated the reading of the parameters but didn't change the default behavior of the instance loading.

Reply:
Is there a benefit or returning a dict? In all our other functions we just return the tuple.
Also I think the _cap_..._reader functions can be removed, as they are one liners that just pass arguments.

Reply:
Checked the current solution:
Is there any benefit of returning a dict? In all other functions we return tuples only.
The one line methods _cap_x_reader seem to just pass arguments. I think we should remove them?

Reply:
I changed the structure of the generic orlib instance loader to also take the formulation function. I think this will give the user the most flexibility and also protect the default behavior in case the user just wants to get the instance ready as defined in orlib.

Reply:
looks good